cv32e40s/tests/programs/custom/pma/pma.c

```diff
@@ -13,319 +13,1147 @@
+
+  // Allocate memory for global pointers here
+  g_has_clic               = calloc(1, sizeof(uint32_t));
+  g_mcause                 = calloc(1, sizeof(mcause_t));
+  g_mepc                   = calloc(1, sizeof(uint32_t));
+  g_mtval                  = calloc(1, sizeof(uint32_t));
+  g_test_num               = calloc(1, sizeof(uint32_t));
+  g_exp_fault              = calloc(1, sizeof(uint32_t));
+  g_recovery_ptr           = calloc(1, sizeof(uint32_t));
+
+  // Setup clic mtvt if clic is enabled
+  *g_has_clic = detect_irq_mode();
+  setup_clic();
+
+  // Add function pointers to new tests here
+  tests[0]  = dummy; // unused, can be used for env sanity checking
+  tests[1]  = exec_only_for_main_regions;
+  tests[2]  = non_natural_aligned_store_to_io;
+  tests[3]  = non_natural_aligned_loads_from_io;
+  tests[4]  = misaligned_fault_nochange_regfile;
+  tests[5]  = misaligned_border_io_to_mem;
+  tests[6]  = misaligned_border_mem_to_io;
+  tests[7]  = misalign_store_fault_no_bus_access_second;
+  tests[8]  = misalign_store_fault_io_no_bus_access_first;
+  // Run all tests in list above
+  cvprintf(V_LOW, "\nPMA test start\n\n");
+  for (volatile uint32_t i = START_TEST_NUM; i < NUM_TESTS; i++) {
+    test_res = set_test_status(tests[i](i, (volatile uint32_t)(0)), test_res);
+  }
 
-// Exception-causing Instructions
+  // Report failures
+  retval = get_result(test_res, tests);
 
-static void (*instr_access_fault)(void) = (void (*)(void))IO_ADDR;
+  // Free dynamically allocated memory
+  free((void *)g_has_clic                  );
+  free((void *)g_mcause                    );
+  free((void *)g_mepc                      );
+  free((void *)g_mtval                     );
+  free((void *)g_test_num                  );
+  free((void *)g_exp_fault                 );
+  free((void *)g_recovery_ptr              );
 
-void misaligned_store(void) {
-  uint32_t tmp;
-  tmp = 0xBBBBBBBB;
-  __asm__ volatile("sw %0, 1(%1)" : "=r"(tmp) : "r"(IO_ADDR));
+  return retval; // Nonzero for failing tests
 }
 
-void load_misaligned_io(void)     {__asm__ volatile("lw t0, 3(%0)" : : "r"(IO_ADDR) : "t0");}
-static void load_aligned_io(void) {__asm__ volatile("lw t0, 0(%0)" : : "r"(IO_ADDR) : "t0");}
-void load_misaligned_io2(void)    {__asm__ volatile("lw t0, 5(%0)" : : "r"(IO_ADDR) : "t0");}
+// -----------------------------------------------------------------------------
 
-void load_misaligned_iomem(void)     {__asm__ volatile("lw t0, 0(%0)" : : "r"(MEM_ADDR_1 - 2) : "t0");}
-static void load_aligned_iomem(void) {__asm__ volatile("lh t0, 0(%0)" : : "r"(MEM_ADDR_1 - 2) : "t0");}
+int cvprintf(volatile verbosity_t verbosity, const char * volatile format, ...){
+  va_list args;
+  volatile int retval = 0;
 
-void load_misaligned_memio(void)     {__asm__ volatile("lw t0, 0(%0)" : : "r"(IO_ADDR - 1) : "t0");}
-static void load_aligned_memio(void) {__asm__ volatile("lb t0, 0(%0)" : : "r"(IO_ADDR - 1) : "t0");}
+  va_start(args, format);
```

Comment:
Would runs without prints be faster if va_start/end be inside the if-statement? (Not that it matters for this small test.)

Reply:
I have a feeling that this might be optimized by the compiler, but I will test and see if it makes a noticeable difference